File: python/src/streaming/live-from-microphone-with-resume.py
import asyncio
import base64
import json

import pyaudio
import requests
from websockets.asyncio.client import ClientConnection, connect
from websockets.exceptions import (
    ConnectionClosed,
    ConnectionClosedError,
)
from helper import (
    print_message,
    InitiateResponse, 
    StreamingConfiguration, 
    get_gladia_key
)

## Constants
GLADIA_API_URL = "https://api.gladia.io"
REGION = "eu-west" # "us-west"

def init_live_session(config: StreamingConfiguration) -> InitiateResponse:
    gladia_key = get_gladia_key()
    response = requests.post(
        f"{GLADIA_API_URL}/v2/live",
        params={"region": REGION},
        headers={"X-Gladia-Key": gladia_key},
        json=config,
        timeout=3,
    )
    if not response.ok:
        print(f"{response.status_code}: {response.text or response.reason}")
        exit(response.status_code)
    return response.json()


async def stop_recording(websocket: ClientConnection) -> None:
    print(">>>>> Ending the recording…")
    await websocket.send(json.dumps({"type": "stop_recording"}))
    await asyncio.sleep(0)


## Sample code
CHANNELS = 1
FORMAT = pyaudio.paInt16
FRAMES_PER_BUFFER = 3200
SAMPLE_RATE = 16_000

STREAMING_CONFIGURATION: StreamingConfiguration = {
    "encoding": "wav/pcm",
    "sample_rate": SAMPLE_RATE,
    "bit_depth": 16,  # It should match the FORMAT value
    "channels": CHANNELS,
    "language_config": {
        "languages": [],
        "code_switching": True,
    },
    "messages_config": {
        "receive_partial_transcripts": False, # Set to True to receive partial/intermediate transcript
        "receive_final_transcripts": True
    }
}

BUFFER = {"data": b"", "bytes_sent": 0}
P = pyaudio.PyAudio()


async def send_audio(stream: pyaudio.Stream, socket: ClientConnection) -> None:
    while True:
        chunk = stream.read(FRAMES_PER_BUFFER)
        BUFFER["data"] += chunk
        data = base64.b64encode(chunk).decode("utf-8")
        json_data = json.dumps({"type": "audio_chunk", "data": {"chunk": str(data)}})
        try:
            await socket.send(json_data)
            await asyncio.sleep(0)
        except ConnectionClosed:
            return


async def receive_messages_from_socket(socket: ClientConnection) -> None:
    try:
        async for message in socket:
            content = json.loads(message)
            if content["type"] == "audio_chunk" and content["acknowledged"]:
                # Use acknowledgement to know when to send the next chunk
                BUFFER["data"] = BUFFER["data"][
                    content["data"]["byte_range"][1] - BUFFER["bytes_sent"] :
                ]
                BUFFER["bytes_sent"] = content["data"]["byte_range"][1]
            print_message(content)
            
    except ConnectionClosedError:
        return


async def force_stop(socket: ClientConnection) -> None:
    await asyncio.sleep(5)
    print(">>>>> Forcibly closing the websocket connection…")
    await socket.close(4500)


async def main():
    response = init_live_session(STREAMING_CONFIGURATION)
    already_stopped = False
    print("\n################ Begin session ################\n")
    stream = P.open(
        format=FORMAT,
        channels=CHANNELS,
        rate=SAMPLE_RATE,
        input=True,
        frames_per_buffer=FRAMES_PER_BUFFER,
    )
    print(">>>>> Starting the recording…")
    async for websocket in connect(response["url"]):
        # Sends the buffered audio data if available on resume
        if len(BUFFER["data"]):
            print(
                f">>>>> Resuming the recording with {len(BUFFER['data'])/1024}KiB in cache…"
            )
            await websocket.send(BUFFER["data"])
            BUFFER["data"] = b""

        tasks = []

        if not already_stopped:
            tasks.append(asyncio.create_task(force_stop(websocket)))
            already_stopped = True

        tasks.append(asyncio.create_task(send_audio(stream, websocket)))
        tasks.append(asyncio.create_task(receive_messages_from_socket(websocket)))
        try:
            await asyncio.wait(tasks)
        except asyncio.exceptions.CancelledError:
            for task in tasks:
                task.cancel()
            await stop_recording(websocket)
            await receive_messages_from_socket(websocket)
            break


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/src/live/live-from-microphone-with-resume.ts
import WebSocket from 'ws';
import {
  getMicrophoneAudioFormat,
  initMicrophoneRecorder,
  printMessage,
  readGladiaApiKey,
} from './helpers.js';
import { InitiateResponse, StreamingConfig } from './types.js';

const gladiaApiKey = readGladiaApiKey();
const gladiaApiUrl = 'https://api.gladia.io';
const region = 'eu-west'; // us-west

const config: StreamingConfig = {
  language_config: {
    languages: [],
    code_switching: false,
  },
  messages_config: {
    receive_partial_transcripts: false, // Set to true to receive partial/intermediate transcript
    receive_final_transcripts: true,
  },
};

async function initLiveSession(): Promise<InitiateResponse> {
  const response = await fetch(`${gladiaApiUrl}/v2/live?region=${region}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-GLADIA-KEY': gladiaApiKey,
    },
    body: JSON.stringify({ ...getMicrophoneAudioFormat(), ...config }),
  });
  if (!response.ok) {
    console.error(
      `${response.status}: ${(await response.text()) || response.statusText}`,
    );
    process.exit(response.status);
  }

  return await response.json();
}

function initWebSocketClient({ url }: InitiateResponse) {
  let socket: WebSocket | null = null;
  let bytesSent = 0;
  let audioBuffer = Buffer.alloc(0);
  let stopRecording = false;

  function initWebSocket() {
    console.log('>>>>> Connecting to websocket');
    socket = new WebSocket(url);

    socket.addEventListener('open', function () {
      console.log('>>>>> Connected to websocket');
      if (audioBuffer.byteLength) {
        socket?.send(audioBuffer);
        audioBuffer = Buffer.alloc(0);
      }
      if (stopRecording) {
        socket?.send(JSON.stringify({ type: 'stop_recording' }));
      }
    });

    socket.addEventListener('error', function (error) {
      console.error(error);
      process.exit(1);
    });

    socket.addEventListener('close', async ({ code, reason }) => {
      if (code === 1000 || stopRecording) {
        process.exit(0);
      } else {
        console.log('>>>>> Lost connection with websocket');
        socket?.removeAllListeners();
        socket = null;
        await new Promise((resolve) => setTimeout(resolve, 1000));
        initWebSocket();
      }
    });

    socket.addEventListener('message', function (event) {
      // All the messages we are sending are in JSON format
      const message = JSON.parse(event.data.toString());
      printMessage(message);

      if (message.type === 'audio_chunk' && message.acknowledged) {
        audioBuffer = audioBuffer.subarray(
          message.data.byte_range[1] - bytesSent,
        );
        bytesSent = message.data.byte_range[1];
      }
    });
  }

  initWebSocket();

  return {
    sendAudioChunk: (chunk: Buffer) => {
      audioBuffer = Buffer.concat([audioBuffer, chunk]);
      if (socket?.readyState === WebSocket.OPEN) {
        socket.send(chunk);
      }
    },
    stopRecording: () => {
      stopRecording = true;
      if (socket?.readyState === WebSocket.OPEN) {
        socket.send(JSON.stringify({ type: 'stop_recording' }));
      }
    },
    forceClose: () => {
      if (socket?.readyState === WebSocket.OPEN) {
        socket.close(4500);
      }
    },
  };
}

async function start() {
  const initiateResponse = await initLiveSession();

  const client = initWebSocketClient(initiateResponse);

  let closeInterval: NodeJS.Timeout | null = null;

  const recorder = initMicrophoneRecorder(
    // Send every chunk from recorder to the socket
    (chunk) => client.sendAudioChunk(chunk),
    // When the recording is stopped, we send a message to tell the server
    // we are done sending audio and it can start the post-processing
    () => {
      if (closeInterval) clearInterval(closeInterval);
      client.stopRecording();
    },
  );

  // We can start the recording without waiting for the connection to be open
  // since the client buffers the audio

  console.log();
  console.log('################ Begin session ################');
  console.log();

  recorder.start();

  closeInterval = setInterval(() => {
    client.forceClose();
  }, 10_000);
}

start();
